File: src/pricing.ts
/**
 * pricing.ts — the only place that knows what a token costs.
 *
 * Resolution order, best first:
 *   1. remote table  (PRICING_URL, fetched + cached in memory, TTL)
 *   2. local file    (PRICING_FILE, default ./pricing.json, hot-reloaded in dev)
 *   3. built-in      (BUILTIN below — conservative, so a broken deploy overcharges
 *                     rather than serving tokens for free)
 *
 * Prices are provider cost only. The margin lives in MARGIN (see billing.ts) and
 * is never part of this table, so the table can be published/shared without
 * leaking what we charge.
 */

import { readFileSync, watch } from "node:fs";
import { resolve as resolvePath } from "node:path";

export interface ModelPrice {
  label?: string;
  /** USD per 1M uncached input tokens (text/image/video). */
  in: number;
  /** USD per 1M output tokens. Includes thinking tokens on Gemini. */
  out: number;
  /** USD per 1M cache-hit input tokens. Falls back to `in` when absent. */
  cached_in?: number;
  /** USD per 1M audio input tokens. Falls back to `in` when absent. */
  audio_in?: number;
  /** Alias whose target Google may repoint at any time. */
  floating?: boolean;
  /** Set when this entry came from `default` rather than an exact match. */
  fallback?: boolean;
  note?: string;
}

export interface PricingTable {
  schema: string;
  version: string;
  source?: string;
  tier?: string;
  currency?: string;
  unit?: string;
  note?: string;
  default: ModelPrice;
  models: Record<string, ModelPrice>;
}

const SCHEMA = "scrambler/pricing@1";

/** Last-resort table. Deliberately expensive: matches the priciest flash tier. */
const BUILTIN: PricingTable = {
  schema: SCHEMA,
  version: "builtin",
  tier: "paid-standard",
  currency: "USD",
  unit: "per_1m_tokens",
  default: { label: "builtin fallback", in: 1.5, out: 9.0, cached_in: 0.15, fallback: true },
  models: {},
};

const FILE = process.env.PRICING_FILE ?? resolvePath(process.cwd(), "pricing.json");
const URL_ = process.env.PRICING_URL ?? "";
const TTL_MS = Number(process.env.PRICING_TTL_SEC ?? 3600) * 1000;
const WATCH = process.env.NODE_ENV !== "production";

let local: PricingTable = BUILTIN;
let remote: PricingTable | null = null;
let remoteEtag: string | null = null;
let remoteFetchedAt = 0;
let remoteInflight: Promise<void> | null = null;

function validate(raw: unknown): PricingTable {
  if (!raw || typeof raw !== "object") throw new Error("pricing: not an object");
  const t = raw as PricingTable;
  if (t.schema !== SCHEMA) throw new Error(`pricing: unexpected schema ${t.schema}`);
  if (!t.default || typeof t.default.in !== "number" || typeof t.default.out !== "number") {
    throw new Error("pricing: missing or invalid default entry");
  }
  if (!t.models || typeof t.models !== "object") throw new Error("pricing: missing models");
  for (const [id, p] of Object.entries(t.models)) {
    if (typeof p?.in !== "number" || typeof p?.out !== "number") {
      throw new Error(`pricing: model ${id} needs numeric in/out`);
    }
    if (p.in < 0 || p.out < 0) throw new Error(`pricing: model ${id} has negative price`);
  }
  return t;
}

function loadLocal(): void {
  try {
    local = validate(JSON.parse(readFileSync(FILE, "utf8")));
    console.log(`[pricing] local table ${local.version} (${Object.keys(local.models).length} models)`);
  } catch (err) {
    console.error(`[pricing] local table unusable, keeping previous/builtin: ${(err as Error).message}`);
  }
}

loadLocal();

if (WATCH) {
  try {
    // Dev convenience: edit pricing.json, next request uses it. No restart.
    watch(FILE, { persistent: false }, () => setTimeout(loadLocal, 50));
  } catch {
    /* file may not exist yet — builtin carries us */
  }
}

async function refreshRemote(): Promise<void> {
  if (!URL_) return;
  if (Date.now() - remoteFetchedAt < TTL_MS) return;
  if (remoteInflight) return remoteInflight;

  remoteInflight = (async () => {
    try {
      const res = await fetch(URL_, {
        headers: remoteEtag ? { "if-none-match": remoteEtag } : {},
        signal: AbortSignal.timeout(5000),
      });
      if (res.status === 304) {
        remoteFetchedAt = Date.now();
        return;
      }
      if (!res.ok) throw new Error(`HTTP ${res.status}`);
      const table = validate(await res.json());
      remote = table;
      remoteEtag = res.headers.get("etag");
      remoteFetchedAt = Date.now();
      console.log(`[pricing] remote table ${table.version}`);
    } catch (err) {
      // Keep the stale remote copy if we have one; otherwise local/builtin serve.
      remoteFetchedAt = Date.now() - TTL_MS / 2; // retry sooner than a full TTL
      console.warn(`[pricing] remote refresh failed: ${(err as Error).message}`);
    } finally {
      remoteInflight = null;
    }
  })();

  return remoteInflight;
}

function table(): PricingTable {
  return remote ?? local;
}

/** Kick off a remote refresh without blocking the request path. */
export function warmPricing(): void {
  void refreshRemote();
}

/**
 * How many days old the active table's `version` date is, or null when the
 * version is not a date.
 *
 * The table is maintained by hand: nothing fetches a provider's real prices,
 * because they publish them as documentation, not as an API. So when a provider
 * raises a rate, this file keeps charging the old one and the operator absorbs
 * the difference — silently, forever. An age is the only cheap signal that
 * something needs re-checking.
 */
export function pricingAgeDays(): number | null {
  const v = table().version;
  const m = /^(\d{4})-(\d{2})-(\d{2})/.exec(v);
  if (!m) return null;
  const then = Date.UTC(Number(m[1]), Number(m[2]) - 1, Number(m[3]));
  return Math.floor((Date.now() - then) / 86_400_000);
}

/** Table identifier that goes into the billing frame, for support/debugging. */
export function pricingVersion(): string {
  const t = table();
  return `${t.version}${remote ? "+remote" : ""}`;
}

/**
 * Price for a model id. Never throws and never returns undefined — an unknown id
 * gets the conservative default with `fallback: true` so it shows up in logs.
 */
/**
 * Does this model have an EXPLICIT price entry (not the fallback)? A model with
 * only the fallback price is a business risk — we might charge less than the
 * provider bills us — so it must never be offered or served. Note an explicit
 * `{in:0, out:0}` (a genuinely free model) counts as priced.
 */
export function hasPrice(model: string): boolean {
  return Object.prototype.hasOwnProperty.call(table().models, model);
}

export function priceFor(model: string): ModelPrice {
  void refreshRemote(); // fire and forget; this call uses whatever is loaded now
  const t = table();
  const hit = t.models[model];
  if (hit) return hit;
  console.warn(`[pricing] no entry for "${model}" — using default`);
  return { ...t.default, fallback: true };
}

/** Everything the /models endpoint needs to show a price hint per model. */
export function priceCatalog(): Array<{ model: string } & ModelPrice> {
  const t = table();
  return Object.entries(t.models).map(([model, p]) => ({ model, ...p }));
}

File: core/src/pricing.rs
// ---------------------------------------------------------------------------
// pricing.rs — the pricing TABLE: per-model provider prices (USD / 1M tokens),
// parsed from pricing.json (schema "scrambler/pricing@1"). The billing MATH lives
// in billing.rs; this just answers "what does model X cost?" and carries the table
// version so a billing frame can say which prices applied.
//
// An unlisted model gets the conservative `default` (fallback = true): we'd rather
// over-charge slightly than under-charge and lose money. Margin is NOT in the table
// — it's applied at billing time from the server's MARGIN env.
// ---------------------------------------------------------------------------

use std::collections::HashMap;

use serde::Deserialize;

use crate::billing::{ModelPrice, Tier};

/// One row of pricing.json (`in`/`out`/`cached_in`/`audio_in` = USD per 1M tokens,
/// `out_text` = text/thinking output rate for image models whose `out` is the
/// image-token rate, `per_image` = USD per generated image, `tier` = "free" |
/// "free-tier" | absent for paid). Unknown fields (label, note, floating, …) are ignored.
#[derive(Deserialize)]
struct RawPrice {
    #[serde(rename = "in")]
    input: f64,
    #[serde(rename = "out")]
    output: f64,
    #[serde(default, rename = "out_text")]
    out_text: Option<f64>,
    #[serde(default, rename = "cached_in")]
    cached_in: Option<f64>,
    #[serde(default, rename = "audio_in")]
    audio_in: Option<f64>,
    #[serde(default)]
    fallback: bool,
    #[serde(default)]
    tier: Option<String>,
    #[serde(default)]
    per_image: Option<f64>,
    /// Human-facing marketing name (e.g. "Nano Banana") — carried alongside the price
    /// so the picker can show it instead of the bare API id. Not part of ModelPrice
    /// (which is Copy); the table keeps it in a separate map.
    #[serde(default)]
    label: Option<String>,
}

impl RawPrice {
    fn to_model_price(&self, force_fallback: bool) -> ModelPrice {
        // Unknown tier strings deliberately land on Paid — mistyping a tier must
        // never accidentally give a model away for free.
        let tier = match self.tier.as_deref() {
            Some("free") => Tier::Free,
            Some("free-tier") => Tier::FreeTier,
            _ => Tier::Paid,
        };
        ModelPrice {
            input: self.input,
            output: self.output,
            output_text: self.out_text,
            cached: self.cached_in,
            audio: self.audio_in,
            fallback: force_fallback || self.fallback,
            tier,
            per_image: self.per_image,
        }
    }
}

#[derive(Deserialize)]
struct RawTable {
    #[serde(default)]
    version: String,
    default: RawPrice,
    models: HashMap<String, RawPrice>,
}

/// The parsed pricing table: a version, a per-model price map, and a fallback default.
pub struct PricingTable {
    version: String,
    default: ModelPrice,
    models: HashMap<String, ModelPrice>,
    labels: HashMap<String, String>,
}

impl PricingTable {
    /// Parse pricing.json. Errors only on malformed JSON / missing `default`.
    pub fn parse(json: &str) -> Result<PricingTable, String> {
        let raw: RawTable = serde_json::from_str(json).map_err(|e| e.to_string())?;
        let models = raw
            .models
            .iter()
            .map(|(k, v)| (k.clone(), v.to_model_price(false)))
            .collect();
        let labels = raw
            .models
            .iter()
            .filter_map(|(k, v)| v.label.clone().map(|l| (k.clone(), l)))
            .collect();
        Ok(PricingTable {
            version: raw.version,
            default: raw.default.to_model_price(true),
            models,
            labels,
        })
    }

    pub fn version(&self) -> &str {
        &self.version
    }

    /// The marketing name for `model` (e.g. "Nano Banana"), if `pricing.json` gives one.
    pub fn label(&self, model: &str) -> Option<&str> {
        self.labels.get(model).map(|s| s.as_str())
    }

    /// Price for `model`, or the conservative default (fallback = true) if unlisted.
    pub fn price(&self, model: &str) -> ModelPrice {
        self.models.get(model).copied().unwrap_or(self.default)
    }
}

// ---------------------------------------------------------------------------
#[cfg(test)]
mod tests {
    use super::*;

    const JSON: &str = r#"{
        "version": "2026-07-30",
        "default": { "in": 1.5, "out": 9, "cached_in": 0.15, "fallback": true },
        "models": {
            "gemini-3.5-flash-lite": { "in": 0.3, "out": 2.5, "cached_in": 0.03, "audio_in": 0.3 },
            "nano": { "in": 0.5, "out": 60, "out_text": 3 }
        }
    }"#;

    #[test]
    fn known_model_is_priced_exactly_and_not_a_fallback() {
        let t = PricingTable::parse(JSON).unwrap();
        assert_eq!(t.version(), "2026-07-30");
        let p = t.price("gemini-3.5-flash-lite");
        assert_eq!(p.input, 0.3);
        assert_eq!(p.output, 2.5);
        assert_eq!(p.cached, Some(0.03));
        assert_eq!(p.audio, Some(0.3));
        assert_eq!(p.output_text, None);
        assert!(!p.fallback);
    }

    #[test]
    fn image_models_carry_a_separate_text_output_rate() {
        let t = PricingTable::parse(JSON).unwrap();
        let p = t.price("nano");
        assert_eq!(p.output, 60.0);
        assert_eq!(p.output_text, Some(3.0));
    }

    #[test]
    fn unlisted_model_falls_back_to_the_conservative_default() {
        let t = PricingTable::parse(JSON).unwrap();
        let p = t.price("some-brand-new-model");
        assert_eq!(p.input, 1.5);
        assert_eq!(p.output, 9.0);
        assert!(p.fallback);
    }
}
